runtime/doc/dev_vimpatch.txt

```diff
@@ -302,4 +302,40 @@ used in new documentation:
 - `{Only when compiled with ...}`: the vast majority of features have been
   made non-optional (see https://github.com/neovim/neovim/wiki/Introduction)
 
+==============================================================================
+FILETYPE DETECTION                 *dev-vimpatch-filetype*
+
+Nvim's filetype detection behavior matches Vim, but is implemented as part of
+|vim.filetype| (see $VIMRUNTIME/lua/vim/filetype.lua).
+
+Pattern matching has several differences:
+- It is done using explicit Lua patterns (without implicit anchoring) instead
+  of Vim regexes: >
+    "*/debian/changelog" -> "/debian/changelog$"
+    "*/bind/db.*" -> "/bind/db%."
+<
+- Filetype patterns are grouped by their parent pattern to improve matching
+  performance. For this to work properly, parent pattern should:
+    - Match at least the same set of strings as filetype patterns inside it.
+      But not too much more.
+    - Be fast to match.
+
+  When adding a new filetype with pattern matching, consider the following:
+  - If there is already a group with appropriate parent pattern, use it.
+  - If there can be a fast and specific enough pattern to group at least
+    3 filetype patterns, add it as a separate grouped entry.
```

Comment:
This seems like a house of cards. What happens if people get this wrong? Is adding a new filetype now 5x harder than it used to be? Is this worth the gain?

Reply:
People can always opt for adding new built-in pattern (which is a quite rare occasion) to ['' parent pattern](https://github.com/neovim/neovim/blob/6d997f8068a89703823f1572c56a6331c9e024aa/runtime/lua/vim/filetype.lua#L2298). This will still work but without any performance gains.

Yes, this is very much worth the 3x-4x gain on common cases and is easily visible when used in real world cases inside plugins. Here is [one example](https://github.com/ibhagwan/fzf-lua/issues/1358#issuecomment-2254174945).